File: arrays-and-strings/check_permutation.py
# 1.2: Check Permutation

# Runtime: O(n) - Space: O(1)
def create_counts(s1: str, s2: str) -> list:
    chars = [0] * 256

    for s in s1:
        chars[ord(s)] += 1

    for s in s2:
        chars[ord(s)] -= 1

    return chars


# Runtime: O(n) - Space: O(1)
def check_permutations(s1: str, s2: str) -> bool:
    if len(s1) != len(s2):
        return False

    counts = create_counts(s1, s2)
    for c in counts:
        if c != 0:
            return False

    return True


print(check_permutations("abc", "cab"))  # True
print(check_permutations("test", "ttew"))  # False
print(check_permutations("geeksforgeeks", "forgeeksgeeks"))  # True

File: 01-arrays-and-strings/javascript/checkPermutation.js
// 1.2: Check Permutation

// Runtime: O(n) - Space: O(1)
function createCounts(s1, s2) {
  const chars = new Array(256).fill(0)

  for (let i = 0; i < s1.length; ++i) {
    chars[s1.charCodeAt(i)] += 1
  }

  for (let i = 0; i < s2.length; ++i) {
    chars[s2.charCodeAt(i)] -= 1
  }

  return chars
}


// Runtime: O(n) - Space: O(1)
function checkPermutations(s1, s2) {
  if (s1.length !== s2.length) return false

  const counts = createCounts(s1, s2)
  for (let i = 0; i < counts.length; ++i) {
    if (counts[i] !== 0) return false
  }

  return true
}


console.log(checkPermutations('abc', 'cab'))  // true
console.log(checkPermutations('test', 'ttew'))  // false
console.log(checkPermutations('geeksforgeeks', 'forgeeksgeeks'))  // true
